url8be3f02398_vnamboo_LearnJava_tgz-pJ8-org_json_TestJ8/src/nio/samples/JServer.java

```diff
@@ -0,0 +1,141 @@
+                        } else if (key.isReadable()) {
+                            // one of our client sockets has received a command and
+                            // we're now ready to read it in
+                            System.out.println("Accepting command!");                            
+                            SocketChannel ch = (SocketChannel)key.channel();
+                            ByteBuffer buf = ByteBuffer.allocate(200);
+                            ch.read(buf);
+                            buf.flip();
+                            Charset charset = Charset.forName("UTF-8");
+                            CharsetDecoder decoder = charset.newDecoder();
+                            CharBuffer cbuf = decoder.decode(buf);
+                            System.out.print(cbuf.toString());
+                            // re-register this socket with the selector, this time
+                            // for writing since we'll want to write something to it
+                            // on the next go-around
+                            ch.register(this.sel, SelectionKey.OP_WRITE);
+                        } else if (key.isWritable()) {
+                            // we are ready to send a response to one of the client sockets
+                            // we had read a command from previously
+                            System.out.println("Sending response!");
+                            SocketChannel ch = (SocketChannel)key.channel();
+                            ch.write(resp);
+                            resp.rewind();
+                            // we may get another command from this guy, so prepare
+                            // to read again. We could also close the channel, but
+                            // that sort of defeats the whole purpose of doing async
+                            ch.register(this.sel, SelectionKey.OP_READ);
+                        }
+                        i.remove();
+                    }
+                }
+            } catch (IOException e) {
+                System.out.println("Error in poll loop");
+                System.out.println(e.getMessage());
+                System.exit(1);
+            }
+        }
+    }
+}
+
+class Acceptor implements Runnable { // inner
+	final ServerSocketChannel serverSocket;
+	public Acceptor(final ServerSocketChannel serverSocket){
+		this.serverSocket=serverSocket;
+	}
+	public void run() {
+		try {
+			SocketChannel c = serverSocket.accept();
+			if (c != null)
+				System.out.println("Here");
+		}catch(IOException ex) { /* ... */ }
+	}
+}
+
+
+public class JServer {
+    public static void main (String[] args) {
+        ServerSocketChannel sch = null;
+        Selector sel = null;
+
+        try {
+            // setup the socket we're listening for connections on.
+            InetSocketAddress addr = new InetSocketAddress(8400);
+            sch = ServerSocketChannel.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *sch*. Currently, there are execution paths that do not contain closure statements, for example, when AbstractSelectableChannel.configureBlocking throws an exception. Ensure *sch* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).